From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Max Filippov <jcmvbkbc@gmail.com>
Date: Sun, 23 Oct 2022 06:06:32 -0700
Subject: [PATCH 03/60] irqchip/irq-xtensa-pic: add support for domain
 hierarchy

Interrupt controllers on xtensa-powered chips are usually cascaded with
core PIC at the top of hierarchy and board-specific controllers
connected to it. Add support for IRQ domain hierarchy to the xtensa PIC
driver to allow expressing this in the device tree.

Signed-off-by: Max Filippov <jcmvbkbc@gmail.com>
(cherry picked from commit fe9cb0a486dd14b14aee0973d607c29e436a6ac4)
---
 drivers/irqchip/irq-xtensa-pic.c | 100 ++++++++++++++++++++++++-------
 1 file changed, 79 insertions(+), 21 deletions(-)

diff --git a/drivers/irqchip/irq-xtensa-pic.c b/drivers/irqchip/irq-xtensa-pic.c
index 44e7be051a2e7..bab2d66f15433 100644
--- a/drivers/irqchip/irq-xtensa-pic.c
+++ b/drivers/irqchip/irq-xtensa-pic.c
@@ -20,27 +20,6 @@
 #include <linux/irqchip/xtensa-pic.h>
 #include <linux/of.h>
 
-/*
- * Device Tree IRQ specifier translation function which works with one or
- * two cell bindings. First cell value maps directly to the hwirq number.
- * Second cell if present specifies whether hwirq number is external (1) or
- * internal (0).
- */
-static int xtensa_pic_irq_domain_xlate(struct irq_domain *d,
-		struct device_node *ctrlr,
-		const u32 *intspec, unsigned int intsize,
-		unsigned long *out_hwirq, unsigned int *out_type)
-{
-	return xtensa_irq_domain_xlate(intspec, intsize,
-			intspec[0], intspec[0],
-			out_hwirq, out_type);
-}
-
-static const struct irq_domain_ops xtensa_irq_domain_ops = {
-	.xlate = xtensa_pic_irq_domain_xlate,
-	.map = xtensa_irq_map,
-};
-
 static void xtensa_irq_mask(struct irq_data *d)
 {
 	u32 irq_mask;
@@ -64,6 +43,10 @@ static void xtensa_irq_ack(struct irq_data *d)
 	xtensa_set_sr(BIT(d->hwirq), intclear);
 }
 
+static void xtensa_irq_eoi(struct irq_data *d)
+{
+}
+
 static int xtensa_irq_retrigger(struct irq_data *d)
 {
 	unsigned int mask = BIT(d->hwirq);
@@ -79,9 +62,84 @@ static struct irq_chip xtensa_irq_chip = {
 	.irq_mask	= xtensa_irq_mask,
 	.irq_unmask	= xtensa_irq_unmask,
 	.irq_ack	= xtensa_irq_ack,
+	.irq_eoi	= xtensa_irq_eoi,
 	.irq_retrigger	= xtensa_irq_retrigger,
 };
 
+/*
+ * Device Tree IRQ specifier translation function which works with one or
+ * two cell bindings. First cell value maps directly to the hwirq number.
+ * Second cell if present specifies whether hwirq number is external (1) or
+ * internal (0).
+ */
+static int xtensa_pic_irq_domain_xlate(struct irq_domain *d,
+				       struct device_node *ctrlr,
+				       const u32 *intspec, unsigned int intsize,
+				       unsigned long *out_hwirq, unsigned int *out_type)
+{
+	return xtensa_irq_domain_xlate(intspec, intsize,
+				       intspec[0], intspec[0],
+				       out_hwirq, out_type);
+}
+
+#ifdef CONFIG_IRQ_DOMAIN_HIERARCHY
+static int xtensa_pic_irq_domain_alloc(struct irq_domain *domain,
+				       unsigned int virq,
+				       unsigned int nr_irqs, void *args)
+{
+	struct irq_fwspec *fwspec = args;
+	irq_hw_number_t hwirq;
+	int rc;
+
+	if (WARN_ON(nr_irqs != 1))
+		return -EINVAL;
+
+	switch (fwspec->param_count) {
+	case 1:
+		hwirq = fwspec->param[0];
+		break;
+	case 2:
+		if (fwspec->param[1] == 0)
+			hwirq = fwspec->param[0];
+		else if (fwspec->param[1] == 1)
+			hwirq = xtensa_map_ext_irq(fwspec->param[0]);
+		else
+			return -EINVAL;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	pr_debug("%s: virq = %d, hwirq = %ld\n", __func__, virq, hwirq);
+	rc = irq_domain_set_hwirq_and_chip(domain, virq, hwirq,
+					   &xtensa_irq_chip, NULL);
+	if (rc)
+		return rc;
+
+	return xtensa_irq_map(domain, virq, hwirq);
+}
+
+static void xtensa_pic_irq_domain_free(struct irq_domain *domain, unsigned int virq,
+				       unsigned int nr_irqs)
+{
+	unsigned int i;
+
+	for (i = 0; i < nr_irqs; i++) {
+		struct irq_data *d = irq_domain_get_irq_data(domain, virq + i);
+		irq_domain_reset_irq_data(d);
+	}
+}
+#endif
+
+static const struct irq_domain_ops xtensa_irq_domain_ops = {
+	.xlate = xtensa_pic_irq_domain_xlate,
+	.map = xtensa_irq_map,
+#ifdef CONFIG_IRQ_DOMAIN_HIERARCHY
+	.alloc = xtensa_pic_irq_domain_alloc,
+	.free = xtensa_pic_irq_domain_free,
+#endif
+};
+
 int __init xtensa_pic_init_legacy(struct device_node *interrupt_parent)
 {
 	struct irq_domain *root_domain =
-- 
2.55.0

